### PR TITLE
If startRecorder fails, the recorder is left in an invalid, unrecoverable state

### PR DESCRIPTION
**Pull request by jbaudanza**
It's common for either `MediaRecorder.start()` or `MediaRecorder.prepare()` to fail if the device doesn't support the requested codec, or there is some problem with the arguments.

Currently, if this happens, the recorder is left in an invalid state:

### `index.ts`
If an error occurs, this._isRecording is still set to `true`, which is incorrect. This PR will catch an error and reset this boolean to `false`

### `RNAudioRecorderPlayerModule.kt`
If an error occurs,  `this.mediaRecorder` will be set to partially or incorrectly initialized object. The next time `startRecord()` is called, this stale object will be reused, which caused a secondary (and more confusing) error.

To fix this, `startRecord` will catch any initialization error and reset `this.mediaRecorder` to null.

### CODE REVIEW

**Comment by isekovanic** [comment]
Just did a bunch of debugging and came to the same conclusion, for the `index.ts` bit at least. It also happens on `iOS` whenever you decline the microphone capability permission and then immediately try recording again (without unmounting whatever's wrapping the recorder so that the internal state gets reset). `this._isRecording` is `true` forever and the permission check is not done again for `this.startRecorder` to fail properly and handled downstream. It is then left in a state where it thinks it's recording and you can't do anything about it, since:

- `this.stopRecorder` fails on the native level (because of [this check](https://github.com/hyochan/react-native-audio-recorder-player/blob/d497c3fbba5634e55a5750384ca2f2865417d380/ios/RNAudioRecorderPlayer.swift#L314), `self.audioRecorder` is `nil` irrecoverably)
- The JS wrapper thinks a recording is in progress and is never going to propagate the promise rejection that's supposed to be returned from `this.startRecorder`

Not sure if this is useful to anyone, but as a workaround I'm doing something along these lines currently:

```
try {  
  // some logic
} catch (err) {
  // some error handling logic
  audioRecorderPlayer._isRecording = false;
  audioRecorderPlayer._hasPausedRecord = false;
}
```

which I'm not particularly a fan of since I'm essentially messing with local state of the class.

Also, I'm pretty certain that [this](https://github.com/hyochan/react-native-audio-recorder-player/blob/d497c3fbba5634e55a5750384ca2f2865417d380/index.ts#L280) and [this](https://github.com/hyochan/react-native-audio-recorder-player/blob/d497c3fbba5634e55a5750384ca2f2865417d380/index.ts#L362) method will also likely produce the same issue if something fails on the native side.

So yeah, huge +1 for this @jbaudanza !

PS: Does it make sense to pull these changes out in a separate PR ? Would be happy to do it if so !

**Comment by jbaudanza** [comment]
> Also, I'm pretty certain that [this](https://github.com/hyochan/react-native-audio-recorder-player/blob/d497c3fbba5634e55a5750384ca2f2865417d380/index.ts#L280) and [this](https://github.com/hyochan/react-native-audio-recorder-player/blob/d497c3fbba5634e55a5750384ca2f2865417d380/index.ts#L362) method will also likely produce the same issue if something fails on the native side.
> 
> So yeah, huge +1 for this @jbaudanza !
> 
> PS: Does it make sense to pull these changes out in a separate PR ? Would be happy to do it if so !

Yeah I agree. Any asynchronous operation should have a fallback to reset the state if there is an error.

Maybe cleaning up the js should be done in a separate PR, since this one already has a lot of native changes.

File: android/src/main/java/com/dooboolab.audiorecorderplayer/RNAudioRecorderPlayerModule.kt [hunk 1]
```diff
@@ -172,7 +192,6 @@ class RNAudioRecorderPlayerModule(private val reactContext: ReactApplicationCont
 
         try {
             mediaRecorder!!.stop()
-            mediaRecorder!!.reset()
```
Comment (hyochan) [inline android/src/main/java/com/dooboolab.audiorecorderplayer/RNAudioRecorderPlayerModule.kt:175]: @jbaudanza Could you kindly check why `reset` should be removed?

**Review by hyochan — APPROVED** [review]
Sorry for the late review!
LGTM!

File: android/src/main/java/com/dooboolab.audiorecorderplayer/RNAudioRecorderPlayerModule.kt (hunk 1 above)
Comment (hyochan) [reply android/src/main/java/com/dooboolab.audiorecorderplayer/RNAudioRecorderPlayerModule.kt:175]: I'll leave this part as remaining question and will merge this `PR`! Thanks @jbaudanza

**Comment by hyochan** [comment]
@isekovanic I'll look into this! Thanks for the comment!

**Comment by efstathiosntonas** [comment]
Hey folks, does anyone noticed if the audio message being recorded sounds robotic or tinny? 

I have several reports from users about it. Downgrading to 3.6.10 fixes the issue so I guess something is probably broken in this PR. 

Find attached 2 audio files. The good file is 174KB while the "bad" is 20KB using the same config,codecs.


[audio_files.zip](https://github.com/user-attachments/files/17174264/audio_files.zip)

**Comment by efstathiosntonas** [comment]
After inspecting the changes in this PR it seems some default values are stripped out if not found from `audioSet`. This does not align with the optional types though, needs to be fixed or add fallback values like it used to be.

The audio quality went back to normal when I added these 2 options in the `audioSet`:

```ts
          AudioEncodingBitRateAndroid: 128000,
          AudioChannelsAndroid: 2,
```

@hyochan what shall we do here? Make the types mandatory to avoid situations like mine or update the `RNAudioRecorderPlayerModule.kt` with fallback values like the old times?

**Comment by jbaudanza** [comment]
@efstathiosntonas Thanks for spotting that. I think setting default fallback values would be a good approach

**Pull request merged**